core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -303,6 +303,292 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
+    static constexpr Iterate inner_direction = Policy::inner_direction;
+    static constexpr int rank                = Policy::rank;
+  };
+
+ public:
+  // V - View
+  template <typename V>
+  ParallelReduce(
+      const FunctorType& f, const Policy& p, const V& v,
+      typename std::enable_if<Kokkos::is_view<V>::value, void*>::type = nullptr)
+      : m_functor(f), m_policy(p), m_space(p.space()), m_result_ptr(v.data()) {}
+
+  ParallelReduce(const FunctorType& f, const Policy& p,
+                 const ReducerType& reducer)
+      : m_functor(f),
+        m_policy(p),
+        m_space(p.space()),
+        m_reducer(reducer),
+        m_result_ptr(reducer.view().data()) {}
+
+ private:
+  template <typename PolicyType, typename Functor, typename Reducer>
+  void sycl_direct_launch(const PolicyType& policy, const Functor& functor,
+                          const Reducer& reducer) const {
+    using ReducerConditional =
+        Kokkos::Impl::if_c<std::is_same<InvalidType, ReducerType>::value,
+                           FunctorType, ReducerType>;
+    using ReducerTypeFwd = typename ReducerConditional::type;
+    using WorkTagFwd =
+        std::conditional_t<std::is_same<InvalidType, ReducerType>::value,
+                           WorkTag, void>;
+    using ValueInit =
+        Kokkos::Impl::FunctorValueInit<ReducerTypeFwd, WorkTagFwd>;
+    using ValueJoin =
+        Kokkos::Impl::FunctorValueJoin<ReducerTypeFwd, WorkTagFwd>;
+    using ValueOps = Kokkos::Impl::FunctorValueOps<FunctorType, WorkTag>;
+
+    // Convenience references
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *m_space.impl_internal_space_instance();
+    sycl::queue& q = *instance.m_queue;
+
+    const int nwork = m_policy.m_num_tiles;
+    const int block_size =
+        std::pow(2, std::ceil(std::log2(m_policy.m_prod_tile_dims)));
+
+    const sycl::range<1> local_range(block_size);
+    // REMEMBER swap local x<->y to be conforming with Cuda/HIP implementation
+    const sycl::range<1> global_range(nwork * block_size);
+    const sycl::nd_range<1> range{global_range, local_range};
+
+    const size_t wgroup_size = range.get_local_range().size();
+    size_t size              = range.get_global_range().size();
+    const auto init_size =
+        std::max<std::size_t>((size + wgroup_size - 1) / wgroup_size, 1);
+    const auto& selected_reducer = ReducerConditional::select(functor, reducer);
+    const unsigned int value_count =
+        FunctorValueTraits<ReducerTypeFwd, WorkTagFwd>::value_count(
+            selected_reducer);
+    const auto results_ptr = static_cast<pointer_type>(sycl::malloc_shared(
+        sizeof(value_type) * std::max(value_count, 1u) * init_size, q));
+    // FIXME_SYCL without this we are running into a race condition
+    const auto results_ptr2 = static_cast<pointer_type>(sycl::malloc_shared(
+        sizeof(value_type) * std::max(value_count, 1u) * init_size, q));
```

Comment:
You forgot to call `sycl::free` on that one.  Use `std::unique_ptr` to prevent such memory leaks to happen in the future.

Reply:
Fixed.